File: android/app/src/main/java/info/bagen/rust/plaoc/microService/ipc/helper/encoding.kt
package info.bagen.rust.plaoc.microService.ipc.helper

import java.net.URLDecoder
import java.net.URLEncoder
import java.util.*

enum class SimpleEncoding(val type:String){
    utf8("utf8"),
    base64("base64")
}

fun simpleEncoder(data: String,encoding: SimpleEncoding): ByteArray {
    val bytes = data.toByteArray(Charsets.UTF_8)
    if (encoding == SimpleEncoding.base64) {
        val encoder: Base64.Encoder = Base64.getEncoder()
        return encoder.encode(bytes)
    }
   return bytes
}

fun simpleDecoder(data: ByteArray,encoding: SimpleEncoding): String{
    if (encoding == SimpleEncoding.base64) {
        val decoder: Base64.Decoder = Base64.getDecoder()
        return decoder.decode(data).toString()
    }
    return  String(data)
}

fun utf8_to_b64 (str: String):String{
    return URLEncoder.encode(str, "utf-8")
};

fun b64_to_utf8(str: String): String {
    return URLDecoder.decode(str, "utf-8")
};

fun dataUrlFromUtf8 (
utf8_string: String,
asBase64: Boolean,
mime: String = ""
): String {
    return if (asBase64) {
        "data:${mime};base64,${b64_to_utf8(utf8_string)}"
    } else {
        "data:${mime};charset=UTF-8,${utf8_to_b64(utf8_string)}";
    }
};

File: next/ts/electron/esm/helper/encoding.js
import { binaryToU8a } from "./binaryHelper.js";
const textEncoder = new TextEncoder();
export const simpleEncoder = (data, encoding) => {
    if (encoding === "base64") {
        const byteCharacters = atob(data);
        const binary = new Uint8Array(byteCharacters.length);
        for (let i = 0; i < byteCharacters.length; i++) {
            binary[i] = byteCharacters.charCodeAt(i);
        }
        return binary;
    }
    return textEncoder.encode(data);
};
const textDecoder = new TextDecoder();
export const simpleDecoder = (data, encoding) => {
    if (encoding === "base64") {
        let binary = "";
        const bytes = binaryToU8a(data);
        for (const byte of bytes) {
            binary += String.fromCharCode(byte);
        }
        return btoa(binary);
    }
    return textDecoder.decode(data);
};
export const utf8_to_b64 = (str) => {
    return btoa(unescape(encodeURIComponent(str)));
};
export const b64_to_utf8 = (str) => {
    return decodeURIComponent(escape(atob(str)));
};
export const dataUrlFromUtf8 = (utf8_string, asBase64, mime = "") => {
    const data_url = asBase64
        ? `data:${mime};base64,${utf8_to_b64(utf8_string)}`
        : `data:${mime};charset=UTF-8,${encodeURIComponent(utf8_string)}`;
    return data_url;
};
